File: specs/pine-to-nautilus-skill/assets/templates/indicator_template.py
"""
${INDICATOR_NAME} Indicator for NautilusTrader

Converted from Pine Script: ${ORIGINAL_PINE_SOURCE}
Original Author: ${ORIGINAL_AUTHOR}

Description:
${DESCRIPTION}

Parameters:
${PARAMETERS_DOC}
"""

from collections import deque
from enum import IntEnum
from typing import Optional

from nautilus_trader.indicators.base.indicator import Indicator
from nautilus_trader.model.data import Bar

# Import sub-indicators as needed
# from nautilus_trader.indicators.average.ema import ExponentialMovingAverage
# from nautilus_trader.indicators.atr import AverageTrueRange


class Signal(IntEnum):
    """Trading signal enumeration."""
    STRONG_SELL = -2
    SELL = -1
    NEUTRAL = 0
    BUY = 1
    STRONG_BUY = 2


class ${CLASS_NAME}(Indicator):
    """
    ${DOCSTRING}
    
    Parameters
    ----------
    period : int
        The lookback period for the indicator.
    ${ADDITIONAL_PARAMS}
    """
    
    def __init__(
        self,
        period: int = 14,
        # Add more parameters here
    ):
        # Initialize parent with parameter list for identification
        super().__init__([period])
        
        # Store parameters
        self.period = period
        
        # Initialize buffers for historical data
        self._closes: deque[float] = deque(maxlen=period)
        self._highs: deque[float] = deque(maxlen=period)
        self._lows: deque[float] = deque(maxlen=period)
        
        # Initialize sub-indicators (if needed)
        # self._ema = ExponentialMovingAverage(period)
        # self._atr = AverageTrueRange(period)
        
        # Previous values for crossover detection
        self._prev_value: float = 0.0
        
        # Output values
        self.value: float = 0.0
        self.signal: Signal = Signal.NEUTRAL
        
        # Additional outputs (uncomment/modify as needed)
        # self.upper_band: float = 0.0
        # self.lower_band: float = 0.0
        # self.trend: int = 0  # 1 = up, -1 = down
    
    def handle_bar(self, bar: Bar) -> None:
        """
        Update the indicator with a new bar.
        
        Parameters
        ----------
        bar : Bar
            The bar to process.
        """
        # Extract values from bar
        close = float(bar.close)
        high = float(bar.high)
        low = float(bar.low)
        # volume = float(bar.volume)  # if needed
        
        # Store previous value for crossover detection
        if self.initialized:
            self._prev_value = self.value
        
        # Add to buffers
        self._closes.append(close)
        self._highs.append(high)
        self._lows.append(low)
        
        # Update sub-indicators
        # self._ema.handle_bar(bar)
        # self._atr.handle_bar(bar)
        
        # Check if we have enough data
        if len(self._closes) < self.period:
            return
        
        # Check if sub-indicators are ready
        # if not self._ema.initialized:
        #     return
        
        # Calculate indicator value
        self.value = self._calculate(close, high, low)
        
        # Generate trading signal
        self.signal = self._generate_signal(close)
        
        # Mark as initialized
        self._set_has_inputs(True)
        self._set_initialized(True)
    
    def _calculate(self, close: float, high: float, low: float) -> float:
        """
        Core calculation logic.
        
        Implement the Pine Script formula here.
        """
        # TODO: Implement calculation logic
        # Example: Simple average
        return sum(self._closes) / len(self._closes)
    
    def _generate_signal(self, close: float) -> Signal:
        """
        Generate trading signal based on indicator state.
        """
        # TODO: Implement signal logic
        # Example: Basic threshold-based signal
        if self.value > 70:
            return Signal.SELL
        elif self.value < 30:
            return Signal.BUY
        return Signal.NEUTRAL
    
    def reset(self) -> None:
        """Reset the indicator to its initial state."""
        self._closes.clear()
        self._highs.clear()
        self._lows.clear()
        
        # Reset sub-indicators
        # self._ema.reset()
        # self._atr.reset()
        
        self._prev_value = 0.0
        self.value = 0.0
        self.signal = Signal.NEUTRAL
        
        self._set_has_inputs(False)
        self._set_initialized(False)
    
    @property
    def name(self) -> str:
        """Return the indicator name for logging."""
        return f"${CLASS_NAME}({self.period})"
    
    # Utility methods for Pine Script equivalents
    
    def _crossover(self, current_a: float, current_b: float, prev_a: float, prev_b: float) -> bool:
        """Pine Script ta.crossover equivalent."""
        return current_a > current_b and prev_a <= prev_b
    
    def _crossunder(self, current_a: float, current_b: float, prev_a: float, prev_b: float) -> bool:
        """Pine Script ta.crossunder equivalent."""
        return current_a < current_b and prev_a >= prev_b
    
    @staticmethod
    def _nz(value: Optional[float], replacement: float = 0.0) -> float:
        """Pine Script nz() equivalent."""
        if value is None:
            return replacement
        try:
            import math
            if math.isnan(value):
                return replacement
        except (TypeError, ValueError):
            pass
        return value

File: specs/pine-to-nautilus-skill/assets/templates/indicator_template.rs
//! ${INDICATOR_NAME} Indicator for NautilusTrader
//!
//! Converted from Pine Script: ${ORIGINAL_PINE_SOURCE}
//! Original Author: ${ORIGINAL_AUTHOR}
//!
//! Description:
//! ${DESCRIPTION}

use std::collections::VecDeque;

use nautilus_indicators::indicator::Indicator;
use nautilus_model::data::bar::Bar;

/// Trading signal enumeration
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i8)]
pub enum Signal {
    StrongSell = -2,
    Sell = -1,
    Neutral = 0,
    Buy = 1,
    StrongBuy = 2,
}

impl Default for Signal {
    fn default() -> Self {
        Signal::Neutral
    }
}

/// ${INDICATOR_NAME} indicator
///
/// ${DOCSTRING}
#[derive(Debug)]
pub struct ${STRUCT_NAME} {
    // Parameters
    period: usize,
    
    // Buffers
    closes: VecDeque<f64>,
    highs: VecDeque<f64>,
    lows: VecDeque<f64>,
    
    // Previous values for crossover detection
    prev_value: f64,
    
    // Output values
    pub value: f64,
    pub signal: Signal,
    
    // State
    initialized: bool,
    has_inputs: bool,
}

impl ${STRUCT_NAME} {
    /// Creates a new instance of ${STRUCT_NAME}
    ///
    /// # Arguments
    ///
    /// * `period` - The lookback period for the indicator
    pub fn new(period: usize) -> Self {
        Self {
            period,
            closes: VecDeque::with_capacity(period),
            highs: VecDeque::with_capacity(period),
            lows: VecDeque::with_capacity(period),
            prev_value: 0.0,
            value: 0.0,
            signal: Signal::default(),
            initialized: false,
            has_inputs: false,
        }
    }
    
    /// Core calculation logic
    fn calculate(&self, close: f64, high: f64, low: f64) -> f64 {
        // TODO: Implement calculation logic
        // Example: Simple average
        if self.closes.is_empty() {
            return 0.0;
        }
        self.closes.iter().sum::<f64>() / self.closes.len() as f64
    }
    
    /// Generate trading signal based on indicator state
    fn generate_signal(&self, close: f64) -> Signal {
        // TODO: Implement signal logic
        if self.value > 70.0 {
            Signal::Sell
        } else if self.value < 30.0 {
            Signal::Buy
        } else {
            Signal::Neutral
        }
    }
    
    // Utility methods for Pine Script equivalents
    
    /// Pine Script ta.crossover equivalent
    #[inline]
    fn crossover(current_a: f64, current_b: f64, prev_a: f64, prev_b: f64) -> bool {
        current_a > current_b && prev_a <= prev_b
    }
    
    /// Pine Script ta.crossunder equivalent
    #[inline]
    fn crossunder(current_a: f64, current_b: f64, prev_a: f64, prev_b: f64) -> bool {
        current_a < current_b && prev_a >= prev_b
    }
    
    /// Pine Script nz() equivalent
    #[inline]
    fn nz(value: f64, replacement: f64) -> f64 {
        if value.is_nan() {
            replacement
        } else {
            value
        }
    }
}

impl Indicator for ${STRUCT_NAME} {
    fn name(&self) -> String {
        format!("${STRUCT_NAME}({})", self.period)
    }
    
    fn has_inputs(&self) -> bool {
        self.has_inputs
    }
    
    fn initialized(&self) -> bool {
        self.initialized
    }
    
    fn handle_bar(&mut self, bar: &Bar) {
        // Extract values from bar
        let close = bar.close.as_f64();
        let high = bar.high.as_f64();
        let low = bar.low.as_f64();
        
        // Store previous value for crossover detection
        if self.initialized {
            self.prev_value = self.value;
        }
        
        // Add to buffers (maintain max size)
        if self.closes.len() >= self.period {
            self.closes.pop_front();
            self.highs.pop_front();
            self.lows.pop_front();
        }
        self.closes.push_back(close);
        self.highs.push_back(high);
        self.lows.push_back(low);
        
        self.has_inputs = true;
        
        // Check if we have enough data
        if self.closes.len() < self.period {
            return;
        }
        
        // Calculate indicator value
        self.value = self.calculate(close, high, low);
        
        // Generate trading signal
        self.signal = self.generate_signal(close);
        
        // Mark as initialized
        self.initialized = true;
    }
    
    fn reset(&mut self) {
        self.closes.clear();
        self.highs.clear();
        self.lows.clear();
        self.prev_value = 0.0;
        self.value = 0.0;
        self.signal = Signal::default();
        self.initialized = false;
        self.has_inputs = false;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_initialization() {
        let indicator = ${STRUCT_NAME}::new(14);
        assert!(!indicator.initialized());
        assert_eq!(indicator.value, 0.0);
    }
    
    #[test]
    fn test_crossover() {
        assert!(${STRUCT_NAME}::crossover(1.0, 0.5, 0.4, 0.5));
        assert!(!${STRUCT_NAME}::crossover(0.4, 0.5, 0.4, 0.5));
    }
    
    #[test]
    fn test_nz() {
        assert_eq!(${STRUCT_NAME}::nz(f64::NAN, 0.0), 0.0);
        assert_eq!(${STRUCT_NAME}::nz(5.0, 0.0), 5.0);
    }
}
